rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -0,0 +1,834 @@
+                        Expr::Literal(ScalarValue::Boolean(r)),
+                    ) => match (l, r) {
+                        (Some(l), Some(r)) => {
+                            Expr::Literal(ScalarValue::Boolean(Some(l != r)))
+                        }
+                        _ => Expr::Literal(ScalarValue::Boolean(None)),
+                    },
+                    (Expr::Literal(ScalarValue::Boolean(b)), _)
+                        if is_boolean_type(&right, schemas) =>
+                    {
+                        match b {
+                            Some(true) => Expr::Not(Box::new(right)),
+                            Some(false) => right,
+                            None => Expr::Literal(ScalarValue::Boolean(None)),
+                        }
+                    }
+                    (_, Expr::Literal(ScalarValue::Boolean(b)))
+                        if is_boolean_type(&left, schemas) =>
+                    {
+                        match b {
+                            Some(true) => Expr::Not(Box::new(left)),
+                            Some(false) => left,
+                            None => Expr::Literal(ScalarValue::Boolean(None)),
+                        }
+                    }
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::NotEq,
+                        right: Box::new(right),
+                    },
+                },
+                _ => Expr::BinaryExpr {
+                    left: Box::new(left),
+                    op: *op,
+                    right: Box::new(right),
+                },
+            }
+        }
+        Expr::Not(expr) => match &**expr {
+            Expr::Not(inner) => optimize_expr(&inner, schemas)?,
+            _ => Expr::Not(Box::new(optimize_expr(&expr, schemas)?)),
+        },
+        Expr::Case {
+            expr,
+            when_then_expr,
+            else_expr,
+        } => {
+            if expr.is_none() {
+                // recurse into CASE WHEN condition expressions
+                Expr::Case {
+                    expr: None,
+                    when_then_expr: when_then_expr
+                        .iter()
+                        .map(|(when, then)| {
+                            Ok((
+                                Box::new(optimize_expr(when, schemas)?),
+                                Box::new(optimize_expr(then, schemas)?),
+                            ))
+                        })
+                        .collect::<Result<_>>()?,
+                    else_expr: match else_expr {
+                        Some(e) => Some(Box::new(optimize_expr(e, schemas)?)),
+                        None => None,
+                    },
+                }
+            } else {
+                // when base expression is specified, when_then_expr conditions are literal values
+                // so we can just skip this case
+                e.clone()
```

Comment:
In general SQL I can't remember if this is always the case (that the WHEN/THEN clause need to be literals) but I think this handling is fine for now

Reply:
This is based on the doc string we have in `logical_plan/expr.rs`:

```
The second form uses a base expression and then a series of "when" clauses that match on a literal value.
``

However, general SQL and spark SQL does support arbitrary expressions in when conditions, so I think it's best for us to match that behavior.